src/impute.jl

```diff
@@ -0,0 +1,157 @@
+
+# Example
+```jldoctest
+julia> using AxisKeys, Impute; using AxisSets: KeyedDataset, flatten;
+
+julia> ds = KeyedDataset(
+           flatten([
+               :train => [
+                   :temp => KeyedArray([1.0 1.1; missing 2.2; 3.0 3.3]; time=1:3, id=[:a, :b]),
+                   :load => KeyedArray([7.0 7.7; 8.0 missing; 9.0 9.9]; time=1:3, loc=[:x, :y]),
+                ],
+                :predict => [
+                   :temp => KeyedArray([1.0 missing; 2.0 2.2; 3.0 3.3]; time=1:3, id=[:a, :b]),
+                   :load => KeyedArray([7.0 7.7; 8.1 missing; 9.0 9.9]; time=1:3, loc=[:x, :y]),
+                ]
+            ])...
+       );
+
+julia> [k => parent(parent(v)) for (k, v) in Impute.filter(ds; dims=:time).data]  # KeyedArray printing isn't consistent in jldoctests
+4-element Array{Pair{Tuple{Symbol,Symbol},Array{Union{Missing, Float64},2}},1}:
+   (:train, :temp) => [3.0 3.3]
+   (:train, :load) => [9.0 9.9]
+ (:predict, :temp) => [3.0 3.3]
+ (:predict, :load) => [9.0 9.9]
+
+julia> [k => parent(parent(v)) for (k, v) in Impute.filter(ds; dims=:loc).data]
+4-element Array{Pair{Tuple{Symbol,Symbol},Array{Union{Missing, Float64},2}},1}:
+   (:train, :temp) => [1.0 1.1; missing 2.2; 3.0 3.3]
+   (:train, :load) => [7.0; 8.0; 9.0]
+ (:predict, :temp) => [1.0 missing; 2.0 2.2; 3.0 3.3]
+ (:predict, :load) => [7.0; 8.1; 9.0]
+```
+"""
+Impute.apply(ds::KeyedDataset, f::Filter; dims) = Impute.apply!(deepcopy(ds), f; dims=dims)
+
+function Impute.apply!(ds::KeyedDataset, f::Filter; dims)
+    # Limit our constraint map to paths containing the supplied dim
+    cmap = filter(constraintmap(ds)) do (constraint, paths)
+        # Because dimpath constraints only really make sense if they end with a shared
+        # dimname we can just filter out constraints that don't end with our desired dimname.
+        constraint.segments[end] === dims
+    end
+
+    # Apply our shared filter mask for each set of constrained paths
+    # NOTE: We're assuming that the constrained paths are mutually exclusive, but in theory
+    # the same component could be processed twice
+    for (constraint, paths) in cmap
+        @debug "$constraint => $paths"
+        # We're assuming this dataset has already been validated so all dimpaths are
+        # already equal
+        mask = trues(length(axiskeys(ds, first(paths))))
+
+        # Pre-extract our component keys and values
+        selection = [p[1:end-1] => ds.data[p[1:end-1]] for p in paths]
+
+        # First pass to determine our shared key mask
+        for (k, v) in selection
+            for (i, s) in enumerate(eachslice(v; dims=dims))
+                mask[i] &= f.func(s)
+            end
+        end
+
+        # Second pass to use selectdim on each component with our mask
+        for (k, v) in selection
+            # copy is so we don't change the data element type to a view
+            ds.data[k] = copy(selectdim(v, NamedDims.dim(dimnames(v), dims), mask))
+        end
+    end
+
```

Comment:
Maybe we should still run `validate` over the dataset before returning?

Reply:
Just realised `Impute.validate` as well. I guess neither are exported but is there any reason one might confuse the two?

Reply:
It's possible, but you're unlikely to see both used in the same code, outside of this package. Most folks will be calling specific functions like `Impute.threshold(...)` and `AxisSets.validate` should be restricted to code that's manually manipulating dataset components. I considered renaming the AxisSets one `_validate`, but I'm not confident yet that folks won't need to call it externally.